packages/astro/src/core/dev/index.ts

```diff
@@ -329,16 +329,25 @@ export class AstroDevServer {
       res.write(html);
       res.end();
     } catch (err: any) {
-      this.viteServer.ws.send({ type: 'error', err });
       const statusCode = 500;
-      const html = serverErrorTemplate({
+      const mod = filePath && this.viteServer.moduleGraph.getModuleById(viteifyURL(filePath));
+      if (mod) {
+        for (const m of [mod, ...mod.importedModules]) {
+          this.viteServer.moduleGraph.invalidateModule(m);
+        }
+      } else {
+        this.viteServer.moduleGraph.invalidateAll();
+      }
+      this.viteServer.ws.send({ type: 'error', err });
+      let html = serverErrorTemplate({
         statusCode,
         title: 'Internal Error',
         tabTitle: '500: Error',
         message: stripAnsi(err.message),
         url: err.url || undefined,
         stack: stripAnsi(err.stack),
       });
+      html = await this.viteServer.transformIndexHtml(pathname, html, pathname);
```

Comment:
This adds HMR to the 500 page so it can self-recover.

IMO adding it to the 404 page on a completely-missing route isn’t necessary (unless a user creates a new `.astro` file; do we want it to auto-refresh?)

Reply:
I think fine to skip for now, but that's an interesting idea!